### ISSUE TITLE
Pipes (for line continuation) break syntax highlighting

### ISSUE THREAD

**Issue by calebmeyer** [ezekg/language-haml#56]
<img width="930" alt="screen shot 2016-04-25 at 9 42 45 am" src="https://cloud.githubusercontent.com/assets/3342777/14787690/1d5da4ba-0aca-11e6-9e31-dcae29d87f1b.png">

I took this screenshot with One Dark Syntax so that we'd have a common test ground. The markup:

``` haml
%span.this#works-fine
%input.uh-oh{               |
  type: 'text',             |
  name: 'oh-no',            |
  title: "what's going on?" |
}                           |
%div.not-highlighted-correctly
  %span.neither-are-children
%a.having-more-brackets{ href: "fixes it" }
%span= "See?"

%span.note {                             |
  this: 'fixes the syntax highlighting'  |
}
%div= "But it will break if you try to render it."
```

It appears that pipe characters break the syntax highlighting. They also are highlighted differently (under my usual syntax theme, photon syntax) sometimes. Randomly they'll be white or purple, and I can't really figure out why. Also if you have an id before the attribute braces in the anchor tag above, it doesn't fix the highlighting (everything after the `#` is comment colored)

**Comment by ezekg** [comment]
Thanks for the repo. You're right, I wasn't aware that the [multi-line syntax requires a trailing pipe](http://haml.info/docs/yardoc/file.FAQ.html#q-multiline). I usually don't use pipes, so that's why I didn't catch it. I'll get a fix out soon.

**Comment by ezekg** [comment]
I haven't forgotten about this, it's just proven a lot harder to fix than I initially thought. Hopefully I'll have some time this weekend to work on this a bit more.

**Comment by calebmeyer** [comment]
No problem. Thanks for your work!

**Comment by Alhadis** [comment]
This is happening because a region of embedded Ruby highlighting isn't terminating at the closing `}` character:

<img src="https://cloud.githubusercontent.com/assets/2346707/26116857/f292902e-3aa7-11e7-9275-3dc85a58617f.png" width="400" alt="Figure 1" />

If you remove the trailing pipe from the line containing the closing brace, everything suddenly highlights fine:

<img src="https://cloud.githubusercontent.com/assets/2346707/26116885/0689444c-3aa8-11e7-94df-4ea74518801f.png" width="400" alt="Figure 2" />

[This line is to blame](https://github.com/ezekg/language-haml/blob/f668b64/grammars/ruby%20haml.cson#L81). I've attached a patch below that fixes this. However, I don't know a single thing about Haml, so I have no idea if these changes are valid syntax or not. I'll leave it up to you.


~~~patch
diff --git a/grammars/ruby haml.cson b/grammars/ruby haml.cson
index d6aab60..a3db55c 100644
--- a/grammars/ruby haml.cson	
+++ b/grammars/ruby haml.cson	
@@ -18,6 +18,9 @@
     ]
   }
   {
+    'include': '#continuation'
+  }
+  {
     'match': '^(!!!)($|\\s.*)'
     'name': 'meta.prolog.haml'
     'captures':
@@ -78,14 +81,14 @@
       }
       {
         'begin': '(?<!\\#)\\{(?=.+(,|(do)|\\{|\\}|\\||(\\#.*))\\s*)'
-        'end': '\\s*\\}(?!\\s*\\,)(?!\\s*\\|)(?!\\#\\{.*\\})'
+        'end': '\\s*\\}'
         'name': 'meta.section.attributes.haml'
         'patterns': [
           {
-            'include': 'source.ruby.rails'
+            'include': '#continuation'
           }
           {
-            'include': '#continuation'
+            'include': 'source.ruby.rails'
           }
           {
             'include': '#rubyline'
@@ -147,10 +150,10 @@
         'name': 'meta.section.object.haml'
         'patterns': [
           {
-            'include': 'source.ruby.rails'
+            'include': '#continuation'
           }
           {
-            'include': '#continuation'
+            'include': 'source.ruby.rails'
           }
           {
             'include': '#rubyline'
@@ -288,7 +291,7 @@
 ]
 'repository':
   'continuation':
-    'match': '(\\|)\\s*\\n'
+    'match': '(\\|)\\s*$\\n?'
     'captures':
       '1':
         'name': 'punctuation.separator.continuation.haml'
~~~~

**Comment by ezekg** [comment]
@Alhadis from my tests, that works great. I'll cut another patch release in a minute. Appreciate the help in getting this cleaned up. 🙂

**Comment by Alhadis** [comment]
Actually, it might be worth waiting until another issue is fixed with comments. [See here](https://github.com/github/linguist/pull/3627#issuecomment-301805747).

**Comment by Alhadis** [comment]
Here's an example of an affected code block:

~~~haml
/ HAML comment
- # ruby comment
string
= "string"
= "string" #comment
= "string #{not_comment}" # comment
~~~

Notice how `#{not_comment}` is being coloured like, well, a comment.

**Comment by ezekg** [comment]
Hmm, it renders fine in Atom. Do you know why it's rendering different on GitHub?

![image](https://cloud.githubusercontent.com/assets/6979737/26118693/6b4b84f8-3a30-11e7-9047-c330491542d5.png)

**Comment by Alhadis** [comment]
Thats currently what I'm waiting on GitHub staff to help with before I can properly investigate. =) The [webapp we use](https://github-lightshow.herokuapp.com/) to preview syntax highlighting is showing the same result as Atom does. This is probably due to outdated dependencies (it's using an ancient version of Linguist to generate previews), so... we'll see once the upgrade is finished.

**Comment by ezekg** [comment]
Gotcha! Sounds good. I'll subscribe that issue so I'll know if I need to do anything, then I can cut a new patch for this issue once that's all figured out.

**Comment by ayaankazerouni** [comment]
OS: Ubuntu 16.04
Atom: 1.17.0
language-haml: 0.24.2

Hi! This seems related, but if it's not, I can create a new issue. I'm seeing syntax highlighting breaking on line continuation as well. But I'm not using the pipe character, I'm breaking to the next line after a comma, which [is allowed according to the Haml docs](http://haml.info/docs/yardoc/file.REFERENCE.html#attributes). Removing the line break after the comma un-breaks the syntax highlighting.

### Broken syntax highlighting:
![screenshot from 2017-05-18 12-38-57](https://cloud.githubusercontent.com/assets/11241828/26213438/0cf5eb44-3bc7-11e7-865a-68b6282d1c7b.png)

### Un-broken syntax highlighting:
![screenshot from 2017-05-18 12-39-25](https://cloud.githubusercontent.com/assets/11241828/26213445/105f89ac-3bc7-11e7-9666-78fa41b6d106.png)

**Comment by ezekg** [comment]
@ayaankazerouni could you try the latest patch version?

**Comment by ayaankazerouni** [comment]
Ok, everything looks good now. Thanks!
